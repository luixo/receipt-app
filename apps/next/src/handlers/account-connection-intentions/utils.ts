import * as trpc from "@trpc/server";

import { Database } from "next-app/db";
import { ACCOUNT_CONNECTIONS_INTENTIONS } from "next-app/db/consts";
import { AccountsId, UsersId } from "next-app/db/models";
import { getUserById } from "next-app/handlers/users/utils";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: string,
	asUserId: UsersId
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select("id")
		.where("email", "=", toEmail)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Account with email ${toEmail} does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId)
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new trpc.TRPCError({
			code: "CONFLICT",
			message: `Account with email ${toEmail} is already connected to user ${connectedUser.name}.`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where("accountId", "=", targetAccount.id)
		.where("targetAccountId", "=", fromAccountId)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where("ownerAccountId", "=", targetAccount.id)
				.where("id", "=", viceVersaIntention.userId)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where("ownerAccountId", "=", fromAccountId)
				.where("id", "=", asUserId)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where("accountId", "=", targetAccount.id)
				.where("targetAccountId", "=", fromAccountId)
				.executeTakeFirst();
		});
		return {
			id: targetAccount.id,
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				created: new Date(),
			})
			.executeTakeFirst();
		return {
			id: targetAccount.id,
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where("accountId", "=", fromAccountId)
				.where("targetAccountId", "=", targetAccount.id)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await getUserById(
				database,
				existingIntention.userId,
				["name"]
			);
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to ${toEmail} as user ${
					existingUser!.name
				}.`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user ${user.name}.`,
			});
		}
		throw e;
	}
};

export const removeIntention = async (
	database: Database,
	intention:
		| {
				targetAccountId: AccountsId;
				accountId: AccountsId;
		  }
		| undefined,
	intentionType: string
) => {
	if (!intention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Intention ${intentionType} does not exist.`,
		});
	}
	await database
		.deleteFrom("accountConnectionsIntentions")
		.where("accountId", "=", intention.accountId)
		.where("targetAccountId", "=", intention.targetAccountId)
		.execute();
};
