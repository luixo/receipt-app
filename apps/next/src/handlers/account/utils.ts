import { Selection } from "kysely";
import { Database, ReceiptsSelectExpression, ReceiptsDatabase } from "../../db";
import { AccountsId } from "../../db/models";

export const getAccountById = <SE extends ReceiptsSelectExpression<"accounts">>(
	database: Database,
	id: AccountsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();
};

export const getAccountByEmail = <
	SE extends ReceiptsSelectExpression<"accounts">
>(
	database: Database,
	email: string,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("email", "=", email)
		.executeTakeFirst();
};
