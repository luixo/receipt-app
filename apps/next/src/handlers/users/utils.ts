import { TRPCError } from "@trpc/server";
import type { SelectQueryBuilder, Selection } from "kysely";

import type { Database, ReceiptsSelectExpression } from "next-app/db";
import type { AccountsId, UsersId } from "next-app/db/models";
import type { ReceiptsDatabase } from "next-app/db/types";

export const getUserById = <SE extends ReceiptsSelectExpression<"users">>(
	database: Database,
	id: UsersId,
	selectExpression: SE[],
	queryBuilder?: <O>(
		qb: SelectQueryBuilder<ReceiptsDatabase, "users", O>,
	) => SelectQueryBuilder<ReceiptsDatabase, "users", O>,
): Promise<Selection<ReceiptsDatabase, "users", SE> | undefined> => {
	let selection = database
		.selectFrom("users")
		.select(selectExpression)
		.where("id", "=", id);
	if (queryBuilder) {
		selection = queryBuilder(selection);
	}
	return selection.executeTakeFirst();
};

export const verifyUsersByIds = async (
	database: Database,
	userIds: UsersId[],
	ownerAccountId: AccountsId,
) => {
	const users = await database
		.selectFrom("users")
		.leftJoin("accounts", (qb) =>
			qb.onRef("accounts.id", "=", "users.connectedAccountId"),
		)
		.select([
			"users.id",
			"users.ownerAccountId",
			"users.name",
			"users.publicName",
			"accounts.id as accountId",
			"accounts.email",
		])
		.where("users.id", "in", userIds)
		.execute();
	if (users.length !== userIds.length) {
		const missedUserIds = userIds.filter(
			(userId) => !users.some((user) => user.id === userId),
		);
		throw new TRPCError({
			code: "NOT_FOUND",
			message: `User(s) ${missedUserIds.join(", ")} do(es) not exist.`,
		});
	}
	const notOwnedUsers = users.filter(
		(user) => user.ownerAccountId !== ownerAccountId,
	);
	if (notOwnedUsers.length !== 0) {
		throw new TRPCError({
			code: "FORBIDDEN",
			message: `Not enough rights to add user(s) ${notOwnedUsers
				.map(({ id }) => id)
				.join(", ")} to a receipt.`,
		});
	}
	return userIds.map((userId) => {
		const { ownerAccountId: disregardedOwnerAccountId, ...user } = users.find(
			({ id }) => id === userId,
		)!;
		return user;
	});
};
