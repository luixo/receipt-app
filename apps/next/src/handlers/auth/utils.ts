import * as trpc from "@trpc/server";

import { DAY } from "app/utils/time";
import type { Database } from "next-app/db";
import type { AccountsId, SessionsSessionId } from "next-app/db/models";
import { generateConfirmEmailEmail } from "next-app/email/utils";
import { getUuid } from "next-app/utils/crypto";
import { getEmailClient, isEmailServiceActive } from "next-app/utils/email";

// How long a session should last
const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long before the session expires should we update the expiration date
const SESSION_SHOULD_UPDATE_DURATION = 28 * DAY;

const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);
export const shouldUpdateExpirationDate = (expirationTimestamp: Date) =>
	expirationTimestamp.valueOf() - Date.now() < SESSION_SHOULD_UPDATE_DURATION;

export const createAuthorizationSession = async (
	database: Database,
	accountId: AccountsId,
) => {
	const uuid = getUuid();
	const expirationDate = getExpirationDate();
	await database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const updateAuthorizationSession = async (
	database: Database,
	authToken: SessionsSessionId,
) => {
	database
		.updateTable("sessions")
		.set({ expirationTimestamp: getExpirationDate() })
		.where("sessionId", "=", authToken)
		.executeTakeFirst();
};

export const removeAuthorizationSession = async (
	database: Database,
	sessionId: SessionsSessionId,
) => {
	await database
		.deleteFrom("sessions")
		.where("sessionId", "=", sessionId)
		.executeTakeFirst();
};

export const sendVerificationEmail = async (email: string, token: string) => {
	if (!isEmailServiceActive()) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient().send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(token),
		});
	} catch (e) {
		throw new trpc.TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
