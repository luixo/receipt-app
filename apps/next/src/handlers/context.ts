import * as trpcNext from "@trpc/server/adapters/next";
import { NextApiRequest, NextApiResponse } from "next";
import { Logger, logger } from "../utils/logger";

export type Context = {
	req: NextApiRequest;
	res: NextApiResponse;
	logger: Logger;
	debug: boolean;
};

export const createContext = (
	opts: trpcNext.CreateNextContextOptions
): Context => ({
	req: opts.req,
	res: opts.res,
	logger,
	debug: Boolean(opts.req.headers.debug),
});
