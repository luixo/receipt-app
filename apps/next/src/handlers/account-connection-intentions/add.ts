import * as trpc from "@trpc/server";
import { z } from "zod";

import { emailSchema } from "app/utils/validation";
import { getDatabase } from "next-app/db";
import { addConnectionIntention } from "next-app/handlers/account-connection-intentions/utils";
import { authProcedure } from "next-app/handlers/trpc";
import { getUserById } from "next-app/handlers/users/utils";
import { userIdSchema } from "next-app/handlers/validation";

export const procedure = authProcedure
	.input(
		z.strictObject({
			userId: userIdSchema,
			email: emailSchema,
		})
	)
	.mutation(async ({ input, ctx }) => {
		const database = getDatabase(ctx);
		const user = await getUserById(database, input.userId, [
			"id",
			"name",
			"connectedAccountId",
			"ownerAccountId",
		]);
		if (!user) {
			throw new trpc.TRPCError({
				code: "NOT_FOUND",
				message: `User ${input.userId} does not exist.`,
			});
		}
		if (user.ownerAccountId !== ctx.auth.accountId) {
			throw new trpc.TRPCError({
				code: "FORBIDDEN",
				message: `User ${input.userId} is not owned by ${ctx.auth.accountId}.`,
			});
		}
		if (user.connectedAccountId) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `User ${input.userId} is already connected to an account.`,
			});
		}
		return addConnectionIntention(
			database,
			ctx.auth.accountId,
			user,
			input.email,
			input.userId
		);
	});
