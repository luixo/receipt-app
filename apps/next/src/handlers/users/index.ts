import * as trpc from "@trpc/server";

import { AuthorizedContext } from "next-app/handlers/context";

import { router as deleteRouter } from "./delete";
import { router as getRouter } from "./get";
import { router as getAvailableRouter } from "./get-available";
import { router as getNameRouter } from "./get-name";
import { router as getPagedRouter } from "./get-paged";
import { router as putRouter } from "./put";
import { router as unlinkRouter } from "./unlink";
import { router as updateRouter } from "./update";

export const router = trpc
	.router<AuthorizedContext>()
	.merge(getRouter)
	.merge(getAvailableRouter)
	.merge(getPagedRouter)
	.merge(putRouter)
	.merge(deleteRouter)
	.merge(updateRouter)
	.merge(getNameRouter)
	.merge(unlinkRouter);
