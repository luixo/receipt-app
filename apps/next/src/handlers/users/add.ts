import { z } from "zod";

import { userNameSchema } from "app/utils/validation";
import type { UsersId } from "next-app/db/models";
import { addConnectionIntention } from "next-app/handlers/account-connection-intentions/utils";
import { authProcedure } from "next-app/handlers/trpc";
import { emailSchema } from "next-app/handlers/validation";

export const procedure = authProcedure
	.input(
		z.strictObject({
			name: userNameSchema,
			publicName: userNameSchema.optional(),
			email: emailSchema.optional(),
		}),
	)
	.mutation(async ({ input, ctx }) => {
		const id: UsersId = ctx.getUuid();
		const { database } = ctx;
		return database.transaction().execute(async () => {
			await database
				.insertInto("users")
				.values({
					id,
					ownerAccountId: ctx.auth.accountId,
					name: input.name,
					publicName: input.publicName,
				})
				.executeTakeFirst();
			if (input.email) {
				const connection = await addConnectionIntention(
					database,
					ctx.auth.accountId,
					{ name: input.name },
					input.email,
					id,
				);
				return { id, connection };
			}
			return { id };
		});
	});
