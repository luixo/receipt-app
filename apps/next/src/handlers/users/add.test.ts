import { faker } from "@faker-js/faker";
import { describe, expect } from "vitest";
import { z } from "zod";

import { createAuthContext } from "@tests/backend/utils/context";
import {
	insertAccount,
	insertAccountConnectionIntention,
	insertAccountWithSession,
	insertConnectedUsers,
	insertUser,
} from "@tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "@tests/backend/utils/expect";
import { test } from "@tests/backend/utils/test";
import { MAX_USERNAME_LENGTH, MIN_USERNAME_LENGTH } from "app/utils/validation";
import { t } from "next-app/handlers/trpc";

import { procedure } from "./add";

const router = t.router({ procedure });

describe("users.add", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			router.createCaller(context).procedure({ name: faker.person.fullName() }),
		);

		describe("email", () => {
			test("invalid", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							name: faker.person.fullName(),
							email: "invalid@@mail.org",
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "email": Invalid email address`,
				);
			});
		});

		(["name", "publicName"] as const).forEach((field) => {
			describe(field, () => {
				test("minimal length", async ({ ctx }) => {
					const { sessionId } = await insertAccountWithSession(ctx);
					const caller = router.createCaller(createAuthContext(ctx, sessionId));
					await expectTRPCError(
						() =>
							caller.procedure({
								name: faker.person.fullName(),
								[field]: "a".repeat(MIN_USERNAME_LENGTH - 1),
							}),
						"BAD_REQUEST",
						`Zod error\n\nAt "${field}": Minimal length for user name is ${MIN_USERNAME_LENGTH}`,
					);
				});

				test("maximum length", async ({ ctx }) => {
					const { sessionId } = await insertAccountWithSession(ctx);
					const caller = router.createCaller(createAuthContext(ctx, sessionId));
					await expectTRPCError(
						() =>
							caller.procedure({
								name: faker.person.fullName(),
								[field]: "a".repeat(MAX_USERNAME_LENGTH + 1),
							}),
						"BAD_REQUEST",
						`Zod error\n\nAt "${field}": Maximum length for user name is ${MAX_USERNAME_LENGTH}`,
					);
				});
			});
		});

		test("target email is not registered", async ({ ctx }) => {
			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const fakeEmail = "non-existent@mail.org";
			await expectTRPCError(
				() =>
					caller.procedure({
						name: faker.person.fullName(),
						email: fakeEmail,
					}),
				"NOT_FOUND",
				`Account with email "${fakeEmail}" does not exist.`,
			);
		});

		describe("email connection intention exceptions", () => {
			test("target email is already connected as another user", async ({
				ctx,
			}) => {
				// Foreign account
				const { id: otherAccountId, email: otherEmail } = await insertAccount(
					ctx,
				);
				// Self account
				const { accountId, sessionId } = await insertAccountWithSession(ctx);
				const [{ name: userName }] = await insertConnectedUsers(ctx, [
					accountId,
					otherAccountId,
				]);
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							name: faker.person.fullName(),
							email: otherEmail,
						}),
					"CONFLICT",
					`Account with email "${otherEmail}" is already connected to user "${userName}".`,
				);
			});

			test("account intention already exists", async ({ ctx }) => {
				// Foreign account
				const { id: otherAccountId, email: otherEmail } = await insertAccount(
					ctx,
				);
				// Self account
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: userId, name: userName } = await insertUser(ctx, accountId);
				// Self account's intention to connect to foreign account
				await insertAccountConnectionIntention(
					ctx,
					accountId,
					otherAccountId,
					userId,
				);

				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							name: faker.person.fullName(),
							email: otherEmail,
						}),
					"CONFLICT",
					`You already has intention to connect to "${otherEmail}" as user "${userName}".`,
				);
			});

			test.skip("user intention already exists", async () => {
				// This will only happen on adding an intention on already existing user
			});
		});
	});

	describe("functionality", () => {
		test("user is added - no public name, no email", async ({ ctx }) => {
			// Verifying other accounts are not affected
			await insertAccount(ctx);

			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = router.createCaller(createAuthContext(ctx, sessionId));

			await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure({ name: faker.person.fullName() }),
			);
		});

		test("user is added - public name, no email", async ({ ctx }) => {
			// Verifying other accounts are not affected
			await insertAccount(ctx);

			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = router.createCaller(createAuthContext(ctx, sessionId));

			await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure({
					name: faker.person.fullName(),
					publicName: faker.person.fullName(),
				}),
			);
		});

		describe("user is added - with email", () => {
			test("has a vice versa intention", async ({ ctx }) => {
				// Foreign account
				const { id: otherAccountId, email: otherEmail } = await insertAccount(
					ctx,
				);
				const { id: otherUserId } = await insertUser(ctx, otherAccountId);
				// Self account
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				// Foreign account's intention to connect to self account
				await insertAccountConnectionIntention(
					ctx,
					otherAccountId,
					accountId,
					otherUserId,
				);

				const asName = faker.person.fullName();
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				const result = await expectDatabaseDiffSnapshot(ctx, () =>
					caller.procedure({
						name: asName,
						email: otherEmail,
					}),
				);
				const uuidSchema = z.string().uuid();
				await expect(uuidSchema.safeParse(result.id).success).toBe(true);
				await expect(result).toStrictEqual<typeof result>({
					connection: {
						connected: true,
						account: {
							id: otherAccountId,
							email: otherEmail,
						},
						user: { name: asName },
					},
					id: result.id,
				});
			});

			test("doesn't have a vice versa intention", async ({ ctx }) => {
				// Foreign account
				const { email: otherEmail, id: otherAccountId } = await insertAccount(
					ctx,
				);
				// Self account
				const { sessionId } = await insertAccountWithSession(ctx);

				const asName = faker.person.fullName();
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				const result = await expectDatabaseDiffSnapshot(ctx, () =>
					caller.procedure({
						name: asName,
						email: otherEmail,
					}),
				);
				const uuidSchema = z.string().uuid();
				await expect(uuidSchema.safeParse(result.id).success).toBe(true);
				await expect(result).toStrictEqual<typeof result>({
					connection: {
						connected: false,
						account: {
							id: otherAccountId,
							email: otherEmail,
						},
						user: { name: asName },
					},
					id: result.id,
				});
			});
		});
	});
});
