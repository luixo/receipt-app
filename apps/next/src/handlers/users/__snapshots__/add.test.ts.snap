// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`users.add > functionality > user is added - no public name, no email 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
+ INSERT INTO public.users VALUES ('5cef6eeb-d06c-414e-813b-705457dfa76a', 'Dr. Jeanne Russel', NULL, '7ee13f62-dbf2-4296-bb95-8154ae2c093a', false, false, NULL);"
`;

exports[`users.add > functionality > user is added - public name, no email 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
+ INSERT INTO public.users VALUES ('3ec84632-ed0b-423d-8191-31544ac6bfc0', 'Helen Rodriguez II', 'Geraldine Friesen III', '165535c3-68d2-4e19-b854-8d166ce9c56f', false, false, NULL);"
`;

exports[`users.add > functionality > user is added - with email > doesn't have a vice versa intention 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
+ INSERT INTO public.users VALUES ('0fb185c3-c540-4168-a8bf-956c8ce3cbb5', 'Gilberto Nolan PhD', NULL, '29c912b8-219f-426a-b3d2-67fca04d4210', false, false, NULL);
+ INSERT INTO public.\\"accountConnectionsIntentions\\" VALUES ('29c912b8-219f-426a-b3d2-67fca04d4210', '0fb185c3-c540-4168-a8bf-956c8ce3cbb5', 'e137d89d-f6d2-4cb1-829e-fa58ed8d7ec1', '2020-01-01 00:00:00');"
`;

exports[`users.add > functionality > user is added - with email > has a vice versa intention 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.users VALUES ('9b2513a6-420a-4961-8f9c-b2ab9d115d3f', 'Mariah', NULL, '2344445d-bb11-4058-b0f9-10ed5ddba150', false, false, NULL);
@@ --- --- @@
- INSERT INTO public.\\"accountConnectionsIntentions\\" VALUES ('2344445d-bb11-4058-b0f9-10ed5ddba150', '9b2513a6-420a-4961-8f9c-b2ab9d115d3f', 'd20a5a4f-9477-4707-8b24-0470962cd596', '2020-01-01 00:00:00');
+ INSERT INTO public.users VALUES ('9b2513a6-420a-4961-8f9c-b2ab9d115d3f', 'Mariah', NULL, '2344445d-bb11-4058-b0f9-10ed5ddba150', false, false, 'd20a5a4f-9477-4707-8b24-0470962cd596');
+ INSERT INTO public.users VALUES ('a51d7aae-0076-49df-8284-1a59ea6590e3', 'Lynette Murray', NULL, 'd20a5a4f-9477-4707-8b24-0470962cd596', false, false, '2344445d-bb11-4058-b0f9-10ed5ddba150');"
`;
